File: Memorize/ReviewViewController.m
//
//  ReviewViewController.m
//  Memorize
//
//  Created by Heather Shelley on 10/2/13.
//  Copyright (c) 2013 Mine. All rights reserved.
//

#import "ReviewViewController.h"
#import "ReviewTabViewController.h"
#import "NSDate+Helpers.h"
#import "Card.h"
#import "UserDataController.h"
#import "CardViewerViewController.h"
#import "UndoObject.h"

@interface ReviewViewController ()

@property (weak, nonatomic) IBOutlet UILabel *questionLabel;
@property (strong, nonatomic) IBOutlet UIButton *questionButton;
@property (weak, nonatomic) IBOutlet UITextView *answerTextView;
@property (strong, nonatomic) IBOutlet UIButton *answerButton;
@property (strong, nonatomic) IBOutlet UIButton *correctButton;
@property (weak, nonatomic) IBOutlet UIButton *undoButton;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *editButton;
@property (strong, nonatomic) IBOutlet UIButton *missedButton;
@property (nonatomic, strong) NSMutableArray *cards;
@property (nonatomic, strong) Card *currentCard;
@property (nonatomic, strong) NSMutableArray *undoStack;

@end

@implementation ReviewViewController


- (void)viewDidLoad {
    self.undoStack = [NSMutableArray array];
    switch (self.reviewType) {
        case ReviewTypeNormal:
            self.cards = [[[UserDataController sharedController] todaysNormalReviewCards] mutableCopy];
            break;
        case ReviewTypeReverse:
            self.cards = [[[UserDataController sharedController] todaysReverseReviewCards] mutableCopy];
            break;
    }
    
    [self updateCurrentCard];
}

- (void)viewWillAppear:(BOOL)animated {
    [self updateUI];
}

- (void)updateUI {
    BOOL hasCurrentCard = self.currentCard ? YES : NO;
    switch (self.reviewType) {
        case ReviewTypeNormal:
            self.title = [NSString stringWithFormat:@"Review (%@)", @(self.cards.count)];
            self.questionButton.enabled = NO;
            self.answerButton.enabled = hasCurrentCard;
            self.answerTextView.hidden = YES;
            self.answerTextView.text = self.currentCard.answer;
            break;
        case ReviewTypeReverse:
            self.title = [NSString stringWithFormat:@"Review Reverse (%@)", @(self.cards.count)];
            self.answerButton.enabled = NO;
            self.questionButton.enabled = hasCurrentCard;
            self.questionLabel.hidden = YES;
            self.answerTextView.text = [[self.currentCard.answer componentsSeparatedByCharactersInSet:[NSCharacterSet decimalDigitCharacterSet]] componentsJoinedByString:@" "];
            break;
    }
    self.questionLabel.text = self.currentCard.question;
    self.editButton.enabled = hasCurrentCard;
    self.correctButton.enabled = NO;
    self.missedButton.enabled = NO;
    self.undoButton.enabled = (self.undoStack.count > 0);
}

- (IBAction)questionPressed:(id)sender {
    self.questionLabel.hidden = NO;
    self.correctButton.enabled = YES;
    self.missedButton.enabled = YES;
}

- (IBAction)answerPressed:(id)sender {
    self.answerTextView.hidden = NO;
    self.correctButton.enabled = YES;
    self.missedButton.enabled = YES;
}

- (IBAction)correctPressed:(id)sender {
    UndoObject *undo = [[UndoObject alloc] init];
    undo.card = [self.currentCard copy];
    undo.correct = YES;
    [self.undoStack addObject:undo];
    [self.currentCard updateCorrectForReviewType:self.reviewType];
    [self.cards removeObject:self.currentCard];
    [self updateCurrentCard];
}

- (IBAction)missedPressed:(id)sender {
    UndoObject *undo = [[UndoObject alloc] init];
    undo.card = [self.currentCard copy];
    undo.correct = NO;
    [self.undoStack addObject:undo];
    [self.currentCard updateMissedForReviewType:self.reviewType];
    [self updateCurrentCard];
}

- (IBAction)undoPressed {
    if (self.undoStack.count > 0) {
        UndoObject *undo = self.undoStack.lastObject;
        [self.undoStack removeObject:undo];
        Card *card = undo.card;
        [card synchronize];
        [self removeCardWithCardID:card.cardID];
        [self.cards addObject:card];
        self.currentCard = card;
        if (undo.correct) {
            switch (self.reviewType) {
                case ReviewTypeNormal:
                    [[UserDataController sharedController] decrementNormalCardsReviewedToday];
                    break;
                case ReviewTypeReverse:
                    [[UserDataController sharedController] decrementReverseCardsReviewedToday];
                    break;
            }
        }
        [self updateUI];
    }
}

- (void)removeCardWithCardID:(NSString *)cardID {
    for (Card *card in self.cards) {
        if ([card.cardID isEqualToString:cardID]) {
            [self.cards removeObject:card];
            break;
        }
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.destinationViewController isKindOfClass:[CardViewerViewController class]]) {
        [segue.destinationViewController setExistingCard:self.currentCard];
    }
}

- (void)updateCurrentCard {
    if (self.cards.count > 0) {
        self.currentCard = self.cards[arc4random_uniform((u_int32_t)self.cards.count)];
    } else {
        self.currentCard = nil;
    }
    
    [self updateUI];
}

@end

File: Memorize/ReviewViewController.swift
//
//  ReviewViewController.swift
//  Memorize
//
//  Created by Heather Shelley on 1/16/17.
//  Copyright © 2017 Mine. All rights reserved.
//

import UIKit

class ReviewViewController: UIViewController {
    var reviewType: ReviewType = .normal
    
    @IBOutlet weak var questionLabel: UILabel?
    @IBOutlet var questionButton: UIButton?
    @IBOutlet weak var answerTextView: UITextView?
    @IBOutlet var answerButton: UIButton?
    @IBOutlet var correctButton: UIButton?
    @IBOutlet weak var undoButton: UIButton?
    @IBOutlet weak var editButton: UIBarButtonItem?
    @IBOutlet var missedButton: UIButton?
    private var cards = [Card]()
    private var currentCard: Card?
    private var undoStack = [UndoObject]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        switch reviewType {
        case .normal:
            cards = UserDataController.shared().todaysNormalReviewCards() as? [Card] ?? []
        case .reverse:
            cards = UserDataController.shared().todaysReverseReviewCards() as? [Card] ?? []
        }
        
        updateCurrentCard()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        updateUI()
    }
    
    func updateUI() {
        let hasCurrentCard: Bool = currentCard != nil ? true : false
        switch reviewType {
        case .normal:
            title = "Review (\(cards.count))"
            questionButton?.isEnabled = false
            answerButton?.isEnabled = hasCurrentCard
            answerTextView?.isHidden = true
            answerTextView?.text = currentCard?.answer
        case .reverse:
            title = "Review Reverse (\(cards.count))"
            answerButton?.isEnabled = false
            questionButton?.isEnabled = hasCurrentCard
            questionLabel?.isHidden = true
            answerTextView?.text = currentCard?.answer.components(separatedBy: CharacterSet.decimalDigits).joined(separator: "")
        }
        
        questionLabel?.text = currentCard?.question
        editButton?.isEnabled = hasCurrentCard
        correctButton?.isEnabled = false
        missedButton?.isEnabled = false
        undoButton?.isEnabled = (undoStack.count > 0)
    }
    
    @IBAction func questionPressed(_ sender: Any) {
        questionLabel?.isHidden = false
        correctButton?.isEnabled = true
        missedButton?.isEnabled = true
    }
    
    @IBAction func answerPressed(_ sender: Any) {
        answerTextView?.isHidden = false
        correctButton?.isEnabled = true
        missedButton?.isEnabled = true
    }
    
    @IBAction func correctPressed(_ sender: Any) {
        guard let currentCard = currentCard else { return }
        
        undoStack.append(UndoObject(card: currentCard, correct: true))
        currentCard.updateCorrect(for: reviewType)
        if let index = cards.index(of: currentCard) {
            cards.remove(at: index)
        }
        updateCurrentCard()
    }
    
    @IBAction func missedPressed(_ sender: Any) {
        guard let currentCard = currentCard else { return }
        
        undoStack.append(UndoObject(card: currentCard, correct: false))
        currentCard.updateMissed(for: reviewType)
        updateCurrentCard()
    }
    
    @IBAction func undoPressed() {
        guard let undo = undoStack.popLast() else { return }
        
        let card = undo.card
        card.synchronize()
        if !cards.contains(where: { $0.cardID == card.cardID }) {
            cards.append(card)
        }
        currentCard = card
        if undo.correct {
            switch reviewType {
            case .normal:
                UserDataController.shared().decrementNormalCardsReviewedToday()
            case .reverse:
                UserDataController.shared().decrementReverseCardsReviewedToday()
            }
        }
        updateUI()
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let cardViewer = segue.destination as? CardViewerViewController else { return }
        cardViewer.existingCard = currentCard
    }
    
    func updateCurrentCard() {
        currentCard = cards.isEmpty ? nil : cards.random()
        self.updateUI()
    }
}
